### DISCUSSION TITLE
Regarding similarity in images

### DISCUSSION THREAD

**Discussion by mreece813** [milvus-io/milvus#16558] [Q&A and General discussion]
Hello I am just starting to use this milvus.io. I am wondering how it works? I also am wondering the algorithm behind comparing the images found in the code. Could someone help explain this in broad terms or even point me in the right direction of where I can find it in the code to see it for myself as well. 
Thanks for the help.

**Comment by yhmo** [comment]
I think this article is a good explanation: https://rom1504.medium.com/image-embeddings-ed1b194d113e

In brief, there are two steps to do image similarity search:
1. Convert images into vectors/embeddings
Each vector is a float number array, if a vector has 128 float numbers, we say it is "128 dimensions".
Usually, we write a vector like this: [1.0, 1.02, 2.45, ...., 3.56]

2. Nearest neighbor search in vectors/embeddings
How to determine two vectors are similar or not? A simple way is calculate their euclidean distance.
Assume there are three 2-dimensional vectors:
a = [1, 2]
b = [1, 3]
c = [4, 5]
Which one("b" or "c") is similar to "a"?
We calculate the euclidean distance between "a" and "b",  (1-1)^2+(2-3)^2 = 1
We calculate the euclidean distance between "a" and "c",  (1-4)^2+(2-5)^2 = 18
Since the distance of "a" "b" is less than distance of "a" "c", we say the "b" is similar to "a".


The Mlvus works for "nearest neighbor search", users need to use other tools to convert images into vectors, input the vectors into milvus, then use milvus to do similarity search.
